File: 788.cpp
struct Solution {
    int rotatedDigits(int n) {
        int count = 0;
        vector<char> memo(n + 1, 0);
        for (int i = 0; i <= n; ++i){
            if (i < 10){
                switch (i){
                    case 0:
                    case 1:
                    case 8:
                        memo[i] = 1;
                        break;
                    case 2:
                    case 5:
                    case 6:
                    case 9:
                        memo[i] = 2;
                        ++count;
                        break;
                }
            } else {
                int a = memo[i / 10], b = memo[i % 10];
                if (a == 1 && b == 1){
                    memo[i] = 1;
                } else if (a != 0 && b != 0) {
                    memo[i] = 2;
                    ++count;
                }
            }
        }
        return count;
    }
};

File: sol/788/788.py
class Solution:
    @staticmethod
    def isRot(s: str) -> bool:
        return all(c in '0125689' for c in s) and any(c in '2569' for c in s)

    val = {0: 0}

    def calc(self, n):
        if n not in self.val:
            self.calc(n - 1)
            self.val[n] = self.val[n - 1] + self.isRot(str(n))

    def rotatedDigits(self, n: int) -> int:
        self.calc(n)
        return self.val[n]
